plugins/redis/redis.cpp

```diff
@@ -408,41 +700,100 @@ Connection * Connection::createConnection(ICodeContext * ctx,  Connection * & _c
             connectionCachingLevelChecked = true;
         }
     }
-
-    if ((connectionCachingLevel && cacheConnections) || connectionCachingLevel == CACHE_ALL_CONNECTIONS)
+    return connectionCachingLevel;
+}
+bool Connection::getCacheSubConnections()
+{
+    if (!cacheSubConnectionsOptChecked)//cacheSubConnectionsOptChecked is std:atomic<bool>. Test to guard against unnecessary critical section
     {
-        if (!_cachedConnection)
+        CriticalBlock block(critsec);
+        if (!cacheSubConnectionsOptChecked)
         {
-            _cachedConnection = new Connection(ctx, options, _database, password, _timeout);
-
-            if (!threadHooked)
+            const char * tmp = getenv("HPCC_REDIS_PLUGIN_CACHE_SUB_CONNECTIONS");
+            //cacheSubConnections is already defaulted to true;
+            if (tmp && *tmp)
             {
-                threadHookChain = addThreadTermFunc(releaseContext);
-                threadHooked = true;
+                //less ops to check on & true however, making cacheSubConnections(false) if not met would do also for misspelled versions.
+                //Since the default is ON, misspelling on or true means/does nothing.
+                if (*tmp == '0' || (stricmp("off", tmp) == 0) || (stricmp("false", tmp) == 0))
+                    cacheSubConnections = false;
             }
+            cacheSubConnectionsOptChecked= true;
+        }
+    }
+    return cacheSubConnections;
+}
+int Connection::getUnsubscribeTimeout()
+{
+    if (!unsubscribeTimeoutChecked)//unsubscribeTimeoutChecked is std:atomic<bool>. Test to guard against unnecessary critical section
+    {
+        CriticalBlock block(critsec);
+        if (!unsubscribeTimeoutChecked)
+        {
+            const char * tmp = getenv("HPCC_REDIS_PLUGIN_UNSUBSCRIBE_TIMEOUT");//ms
+            if (tmp && *tmp)
+                unsubscribeTimeout = atoi(tmp);
+            unsubscribeTimeoutChecked = true;
+        }
+    }
+    return unsubscribeTimeout;
+}
+bool Connection::canCacheConnections(bool cachedConnectionRequested, bool isSubscription)
+{
+#if HIREDIS_VERSION_OK_FOR_CACHING
+    switch (getConnectionCachingLevel())
+    {
+    case CACHE_ALL_CONNECTIONS :
+        return true;
+    case NO_CONNECTION_CACHING :
+        return false;
+    }
+
+    if (isSubscription)
+        return cachedConnectionRequested && getCacheSubConnections();
+
+    return cachedConnectionRequested;
+#endif
+    return false;
+}
+static void addThreadHook()
+{
+    if (!threadHooked)
+    {
+        threadHookChain = addThreadTermFunc(releaseAllCachedContexts);
+        threadHooked = true;
+    }
+}
+Connection * Connection::createConnection(ICodeContext * ctx,  Connection * & _cachedConnection, const char * _options, const char * _ip, int _port, bool parseOptions, int _database, const char * password, unsigned _timeout, bool cachedConnectionRequested, bool isSubscription)
```

Comment:
So I've been doing this since I started. The idea is to make it obvious/explicit that these are neither the global nor member versions. This is why not all of the parameters (in this example) are underscored. What is are style alternative?